{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 10 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "/Users/gaoyueya/miniconda3/envs/T2T-ACE-env/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import T2T_ACE.duplication_evaluation as duplication_evaluation\n",
    "from T2T_ACE.alignment_utilities import load_reference\n",
    "import T2T_ACE.validator as v\n",
    "import importlib\n",
    "import T2T_ACE.dup_basepair_correction as dup_basepair_correction"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-14T20:00:34.661470Z",
     "start_time": "2023-12-14T20:00:33.661954Z"
    }
   },
   "id": "94226748497d"
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Loading reference from: /Users/gaoyueya/Documents/Reference_Genome/hg002v1.0.fasta\n",
      "INFO:root:Loading reference from: /Users/gaoyueya/Documents/Reference_Genome/Homo_sapiens_assembly38.fasta\n"
     ]
    }
   ],
   "source": [
    "# Update the reference genomes to local path\n",
    "class ReferenceLocations:\n",
    "        hg002t2t = \"/Users/gaoyueya/Documents/Reference_Genome/hg002v1.0.fasta\"\n",
    "        chm13 = \"/Users/gaoyueya/Documents/Reference_Genome/chm13v2.0.fa.gz\"\n",
    "        hg38 = \"/Users/gaoyueya/Documents/Reference_Genome/Homo_sapiens_assembly38.fasta\"\n",
    "\n",
    "# Load the minimap2 aligner from reference fasta file\n",
    "# Load HG002 T2T reference\n",
    "hg002t2t = load_reference(ReferenceLocations.hg002t2t)\n",
    "# Load hg38 reference\n",
    "hg38 = load_reference(ReferenceLocations.hg38)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-14T20:02:45.425921Z",
     "start_time": "2023-12-14T20:00:34.662775Z"
    }
   },
   "id": "17022b1e2ad1f9f2"
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "outputs": [
    {
     "data": {
      "text/plain": "<module 'T2T_ACE.validator' from '/Users/gaoyueya/Documents/GitHub/T2T-ACE/T2T_ACE/validator.py'>"
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(v)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:13:36.907101Z",
     "start_time": "2023-12-15T02:13:36.899713Z"
    }
   },
   "id": "9f12c3fa9e05a104"
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "outputs": [
    {
     "data": {
      "text/plain": "<module 'T2T_ACE.dup_basepair_correction' from '/Users/gaoyueya/Documents/GitHub/T2T-ACE/T2T_ACE/dup_basepair_correction.py'>"
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(dup_basepair_correction)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:13:37.285676Z",
     "start_time": "2023-12-15T02:13:37.274634Z"
    }
   },
   "id": "e92723087a36e0d0"
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "outputs": [
    {
     "data": {
      "text/plain": "<module 'T2T_ACE.duplication_evaluation' from '/Users/gaoyueya/Documents/GitHub/T2T-ACE/T2T_ACE/duplication_evaluation.py'>"
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "importlib.reload(duplication_evaluation)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:13:37.656610Z",
     "start_time": "2023-12-15T02:13:37.633563Z"
    }
   },
   "id": "57b50d5377ba45a6"
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "outputs": [],
   "source": [
    "# BasePair Correction Dev\n",
    "#TODO: Check the alignment in hg38 and hg2 and obtain the difference between the real DUP event vs DRAGEN call DONE\n",
    "#TODO: Complete the extension function DONE\n",
    "#TODO: Incorporate those functions to dup_eval class DONE\n",
    "#TODO: Added the alignment results of the corrected interval"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-13T16:26:17.152631Z",
     "start_time": "2023-12-13T16:26:17.148432Z"
    }
   },
   "id": "fe50678b769c5c7d"
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr6:256330-381943\n",
      "hg38 dup count: 1\n",
      "interval: chr6:256329-381943\tstrand: 1, start: 0, end: 125614\n",
      "hg2 dup count: 3\n",
      "interval: chr6_MATERNAL:209000-339723\tstrand: 1, start: 0, end: 125614\n",
      "interval: chr16_MATERNAL:35170607-35297731\tstrand: -1, start: 177, end: 125614\n",
      "interval: chr6_PATERNAL:209734-338805\tstrand: 1, start: 0, end: 125614\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chr6:256330-381943(125614bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be moved to the right by 177bp\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be checked by extend_2_right\n",
      "**Using extend_2_right to check the end**\n",
      "extending window size 12562\n",
      "new interval chr6:256330-394506 138177\n",
      "[[126132, 1], [138177, 2]]\n",
      "extend end to the right by 517bp\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chr6:256330-381943 (125614bp)\n",
      "The corrected interval is chr6:256507-382462 (125956bp)\n",
      "The corrected interval is 342bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr6:256507-382462\n",
      "hg38 dup count: 1\n",
      "interval: chr6:256506-382462\tstrand: 1, start: 0, end: 125956\n",
      "hg2 dup count: 3\n",
      "interval: chr6_MATERNAL:209177-340242\tstrand: 1, start: 0, end: 125956\n",
      "interval: chr16_MATERNAL:35170089-35297731\tstrand: -1, start: 0, end: 125955\n",
      "interval: chr6_PATERNAL:209910-339324\tstrand: 1, start: 0, end: 125956\n",
      "----------------------------------------------------------------\n",
      "End of Analysis of DUP interval chr6:256330-381943(125614bp)\n",
      "----------------------------------------------------------------\n"
     ]
    },
    {
     "data": {
      "text/plain": "{'dup_interval': 'chr6:256330-381943',\n 'dup_interval_size': 125614,\n 'original_hg38_hit_count': 1,\n 'original_hg2_hit_count': 3,\n 'original_hg2_mat_hit_count': 2,\n 'original_hg2_pat_hit_count': 1,\n 'original_dup_interval_major_classification': 'Duplication',\n 'original_dup_interval_sub_classification': 'Maternal Heterozygous Duplication',\n 'original_dup_interval_contain_big_gap': False,\n 'corrected_interval': 'chr6:256507-382462',\n 'corrected_interval_size': 125956,\n 'expanded_length': 342,\n 'corrected_hg38_hit_count': 1,\n 'corrected_hg2_hit_count': 3,\n 'corrected_hg2_mat_hit_count': 2,\n 'corrected_hg2_pat_hit_count': 1,\n 'corrected_dup_interval_major_classification': 'Duplication',\n 'corrected_dup_interval_sub_classification': 'Maternal Heterozygous Duplication',\n 'corrected_pos_accuracy': 1,\n 'corrected_end_accuracy': 1}"
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example 1: Shrink and Extend\n",
    "duplication_evaluation.eval_dup_interval(\"chr6:256330-381943\", ReferenceLocations.hg38, ReferenceLocations.hg002t2t, hg38, hg002t2t).analyzeDuplicationIntervals()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:14:24.286016Z",
     "start_time": "2023-12-15T02:14:20.828069Z"
    }
   },
   "id": "ed7401831205ee9a"
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr14:105616092-105632532\n",
      "hg38 dup count: 1\n",
      "interval: chr14:105616091-105632532\tstrand: 1, start: 0, end: 16441\n",
      "hg2 dup count: 4\n",
      "interval: chr14_MATERNAL:107615085-107631645\tstrand: 1, start: 0, end: 16441\n",
      "interval: chr14_MATERNAL:107634637-107651251\tstrand: 1, start: 0, end: 16441\n",
      "interval: chr14_PATERNAL:105325487-105342047\tstrand: 1, start: 0, end: 16441\n",
      "interval: chr14_PATERNAL:105345039-105361644\tstrand: 1, start: 0, end: 16441\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chr14:105616092-105632532(16441bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be checked by extend_2_left\n",
      "**Using extend_2_left to check the pos**\n",
      "extending window size 1644\n",
      "new interval chr14:105614448-105632532 18085\n",
      "[[1, 4]]\n",
      "1 100.0 %\n",
      "new interval chr14:105612804-105632532 19729\n",
      "[[0, 2], [173, 2]]\n",
      "extend pos to the left by 3115bp\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be checked by extend_2_right\n",
      "**Using extend_2_right to check the end**\n",
      "extending window size 1644\n",
      "new interval chr14:105616092-105634177 18086\n",
      "[[18086, 4]]\n",
      "18086 100.0 %\n",
      "\n",
      "new interval chr14:105616092-105635821 19730\n",
      "[[19730, 4]]\n",
      "19730 100.0 %\n",
      "new interval chr14:105616092-105637465 21374\n",
      "[[21374, 4]]\n",
      "21374 100.0 %\n",
      "\n",
      "new interval chr14:105616092-105639109 23018\n",
      "[[23018, 4]]\n",
      "23018 100.0 %\n",
      "\n",
      "new interval chr14:105616092-105640753 24662\n",
      "[[24662, 4]]\n",
      "24662 100.0 %\n",
      "new interval chr14:105616092-105642397 26306\n",
      "[[26306, 4]]\n",
      "26306 100.0 %\n",
      "\n",
      "new interval chr14:105616092-105644041 27950\n",
      "[[27950, 4]]\n",
      "27950 100.0 %\n",
      "new interval chr14:105616092-105645685 29594\n",
      "[[29594, 4]]\n",
      "29594 100.0 %\n",
      "\n",
      "new interval chr14:105616092-105647329 31238\n",
      "[[31238, 4]]\n",
      "31238 100.0 %\n",
      "new interval chr14:105616092-105648973 32882\n",
      "[[32882, 4]]\n",
      "32882 100.0 %\n",
      "\n",
      "new interval chr14:105616092-105650617 34526\n",
      "[[33820, 2], [34526, 2]]\n",
      "extend end to the right by 17378bp\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chr14:105616092-105632532 (16441bp)\n",
      "The corrected interval is chr14:105612977-105649912 (36936bp)\n",
      "The corrected interval is 20495bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr14:105612977-105649912\n",
      "hg38 dup count: 1\n",
      "interval: chr14:105612976-105649912\tstrand: 1, start: 0, end: 36936\n",
      "hg2 dup count: 4\n",
      "interval: chr14_MATERNAL:107631546-107668913\tstrand: 1, start: 0, end: 36936\n",
      "interval: chr14_MATERNAL:107611975-107649785\tstrand: 1, start: 0, end: 36935\n",
      "interval: chr14_PATERNAL:105341948-105379306\tstrand: 1, start: 0, end: 36936\n",
      "interval: chr14_PATERNAL:105322377-105360178\tstrand: 1, start: 0, end: 36935\n",
      "----------------------------------------------------------------\n",
      "End of Analysis of DUP interval chr14:105616092-105632532(16441bp)\n",
      "----------------------------------------------------------------\n"
     ]
    },
    {
     "data": {
      "text/plain": "{'dup_interval': 'chr14:105616092-105632532',\n 'dup_interval_size': 16441,\n 'original_hg38_hit_count': 1,\n 'original_hg2_hit_count': 4,\n 'original_hg2_mat_hit_count': 2,\n 'original_hg2_pat_hit_count': 2,\n 'original_dup_interval_major_classification': 'Duplication',\n 'original_dup_interval_sub_classification': 'Homozygous Duplication',\n 'original_dup_interval_contain_big_gap': False,\n 'corrected_interval': 'chr14:105612977-105649912',\n 'corrected_interval_size': 36936,\n 'expanded_length': 20495,\n 'corrected_hg38_hit_count': 1,\n 'corrected_hg2_hit_count': 4,\n 'corrected_hg2_mat_hit_count': 2,\n 'corrected_hg2_pat_hit_count': 2,\n 'corrected_dup_interval_major_classification': 'Duplication',\n 'corrected_dup_interval_sub_classification': 'Homozygous Duplication',\n 'corrected_pos_accuracy': 1,\n 'corrected_end_accuracy': 1}"
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example 2: Extend on both ends\n",
    "duplication_evaluation.eval_dup_interval(\"chr14:105616092-105632532\", ReferenceLocations.hg38, ReferenceLocations.hg002t2t, hg38, hg002t2t).analyzeDuplicationIntervals()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:14:37.625519Z",
     "start_time": "2023-12-15T02:14:35.497043Z"
    }
   },
   "id": "398968325e21850f"
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr10:39364453-39376272\n",
      "hg38 dup count: 1\n",
      "interval: chr10:39364452-39376272\tstrand: 1, start: 0, end: 11820\n",
      "hg2 dup count: 6\n",
      "interval: chr10_MATERNAL:39471150-39482958\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_MATERNAL:39533724-39545544\tstrand: 1, start: 0, end: 11820\n",
      "interval: chr10_MATERNAL:39523295-39535103\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_MATERNAL:39481579-39493387\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_MATERNAL:39492008-39503816\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_PATERNAL:39517717-39529534\tstrand: 1, start: 0, end: 11820\n",
      "input dup interval: chr10:39364453-39376272\n",
      "hg38 dup count: 1\n",
      "interval: chr10:39364452-39376272\tstrand: 1, start: 0, end: 11820\n",
      "hg2 dup count: 6\n",
      "interval: chr10_MATERNAL:39471150-39482958\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_MATERNAL:39533724-39545544\tstrand: 1, start: 0, end: 11820\n",
      "interval: chr10_MATERNAL:39523295-39535103\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_MATERNAL:39481579-39493387\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_MATERNAL:39492008-39503816\tstrand: 1, start: 0, end: 11815\n",
      "interval: chr10_PATERNAL:39517717-39529534\tstrand: 1, start: 0, end: 11820\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chr10:39364453-39376272(11820bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be checked by extend_2_left\n",
      "**Using extend_2_left to check the pos**\n",
      "extending window size 1182\n",
      "new interval chr10:39363271-39376272 13002\n",
      "[[0, 3]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr10:39362089-39376272 14184\n",
      "[[0, 2]]\n",
      "the window size is too big\n",
      "Keep the last iteration pos which means extend pos to the left 1182bp\n",
      "extending window size 1182\n",
      "new interval chr10:39363271-39376272 13002\n",
      "[[0, 3]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr10:39362089-39376272 14184\n",
      "[[0, 2]]\n",
      "the window size is too big\n",
      "Keep the last iteration pos which means extend pos to the left 1182bp\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be moved to the left by 5bp\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chr10:39364453-39376272 (11820bp)\n",
      "The corrected interval is chr10:39363271-39376268 (12998bp)\n",
      "The corrected interval is 1178bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr10:39363271-39376268\n",
      "hg38 dup count: 1\n",
      "interval: chr10:39363270-39376268\tstrand: 1, start: 0, end: 12998\n",
      "hg2 dup count: 3\n",
      "interval: chr10_MATERNAL:39469967-39482958\tstrand: 1, start: 0, end: 12997\n",
      "interval: chr10_MATERNAL:39532505-39545540\tstrand: 1, start: 0, end: 12998\n",
      "interval: chr10_PATERNAL:39516534-39529530\tstrand: 1, start: 0, end: 12998\n",
      "input dup interval: chr10:39363271-39376268\n",
      "hg38 dup count: 1\n",
      "interval: chr10:39363270-39376268\tstrand: 1, start: 0, end: 12998\n",
      "hg2 dup count: 3\n",
      "interval: chr10_MATERNAL:39469967-39482958\tstrand: 1, start: 0, end: 12997\n",
      "interval: chr10_MATERNAL:39532505-39545540\tstrand: 1, start: 0, end: 12998\n",
      "interval: chr10_PATERNAL:39516534-39529530\tstrand: 1, start: 0, end: 12998\n"
     ]
    },
    {
     "data": {
      "text/plain": "{'dup_interval': 'chr10:39364453-39376272',\n 'dup_interval_size': 11820,\n 'original_hg38_hit_count': 1,\n 'original_hg2_hit_count': 6,\n 'original_hg2_mat_hit_count': 5,\n 'original_hg2_pat_hit_count': 1,\n 'dup_interval_classification': 'Maternal Heterozygous Duplication',\n 'corrected_interval': 'chr10:39363271-39376268',\n 'corrected_interval_size': 12998,\n 'expanded_length': 1178,\n 'corrected_hg38_hit_count': 1,\n 'corrected_hg2_hit_count': 3,\n 'corrected_hg2_mat_hit_count': 2,\n 'corrected_hg2_pat_hit_count': 1,\n 'corrected_dup_interval_classification': 'Maternal Heterozygous Duplication',\n 'corrected_pos_accuracy': 0,\n 'corrected_end_accuracy': 1}"
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example 3: DUP that can't be precisely corrected\n",
    "duplication_evaluation.eval_dup_interval(\"chr10:39364453-39376272\", ReferenceLocations.hg38, ReferenceLocations.hg002t2t, hg38, hg002t2t).analyzeDuplicationIntervals()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-13T14:12:52.032578Z",
     "start_time": "2023-12-13T14:12:50.466522Z"
    }
   },
   "id": "d0a30766c93f1c2e"
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chrX:49017567-49018675\n",
      "hg38 dup count: 1\n",
      "interval: chrX:49017566-49018675\tstrand: 1, start: 0, end: 1109\n",
      "hg2 dup count: 2\n",
      "interval: chrX_MATERNAL:48440458-48441566\tstrand: 1, start: 0, end: 1109\n",
      "interval: chrX_MATERNAL:48441566-48442674\tstrand: 1, start: 1, end: 1109\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chrX:49017567-49018675(1109bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be checked by extend_2_left\n",
      "**Using extend_2_left to check the pos**\n",
      "Basepair correction Method doesn't support DUPs on chrX or chrY\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be checked by extend_2_right\n",
      "**Using extend_2_right to check the end**\n",
      "Basepair correction Method doesn't support DUPs on chrX or chrY\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chrX:49017567-49018675 (1109bp)\n",
      "The corrected interval is chrX:49017567-49018676 (1110bp)\n",
      "The corrected interval is 1bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chrX:49017567-49018676\n",
      "hg38 dup count: 1\n",
      "interval: chrX:49017566-49018676\tstrand: 1, start: 0, end: 1110\n",
      "hg2 dup count: 2\n",
      "interval: chrX_MATERNAL:48441566-48442675\tstrand: 1, start: 1, end: 1110\n",
      "interval: chrX_MATERNAL:48440458-48441567\tstrand: 1, start: 0, end: 1110\n",
      "----------------------------------------------------------------\n",
      "End of Analysis of DUP interval chrX:49017567-49018675(1109bp)\n",
      "----------------------------------------------------------------\n"
     ]
    },
    {
     "data": {
      "text/plain": "{'dup_interval': 'chrX:49017567-49018675',\n 'dup_interval_size': 1109,\n 'original_hg38_hit_count': 1,\n 'original_hg2_hit_count': 2,\n 'original_hg2_mat_hit_count': 2,\n 'original_hg2_pat_hit_count': 0,\n 'original_dup_interval_major_classification': 'Duplication',\n 'original_dup_interval_sub_classification': 'Maternal Heterozygous Duplication',\n 'original_dup_interval_contain_big_gap': False,\n 'corrected_interval': 'chrX:49017567-49018676',\n 'corrected_interval_size': 1110,\n 'expanded_length': 1,\n 'corrected_hg38_hit_count': 1,\n 'corrected_hg2_hit_count': 2,\n 'corrected_hg2_mat_hit_count': 2,\n 'corrected_hg2_pat_hit_count': 0,\n 'corrected_dup_interval_major_classification': 'Duplication',\n 'corrected_dup_interval_sub_classification': 'Maternal Heterozygous Duplication',\n 'corrected_pos_accuracy': 0,\n 'corrected_end_accuracy': 0}"
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example 4: DUP on the sex chromosome\n",
    "duplication_evaluation.eval_dup_interval(\"chrX:49017567-49018675\", ReferenceLocations.hg38, ReferenceLocations.hg002t2t, hg38, hg002t2t).analyzeDuplicationIntervals()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:17:31.599046Z",
     "start_time": "2023-12-15T02:17:31.555435Z"
    }
   },
   "id": "e85427494c5c7747"
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr10:46514225-46561077\n",
      "hg38 dup count: 1\n",
      "interval: chr10:46514224-46561077\tstrand: 1, start: 0, end: 46853\n",
      "hg2 dup count: 5\n",
      "interval: chr10_MATERNAL:47843348-47890485\tstrand: 1, start: 0, end: 46853\n",
      "interval: chr10_MATERNAL:48513086-48560225\tstrand: 1, start: 0, end: 46853\n",
      "interval: chr10_MATERNAL:49784540-49831705\tstrand: -1, start: 0, end: 46853\n",
      "interval: chr10_PATERNAL:48447360-48494479\tstrand: 1, start: 0, end: 46853\n",
      "interval: chr10_PATERNAL:49717006-49764147\tstrand: -1, start: 0, end: 46853\n",
      "input dup interval: chr10:46514225-46561077\n",
      "hg38 dup count: 1\n",
      "interval: chr10:46514224-46561077\tstrand: 1, start: 0, end: 46853\n",
      "hg2 dup count: 5\n",
      "interval: chr10_MATERNAL:47843348-47890485\tstrand: 1, start: 0, end: 46853\n",
      "interval: chr10_MATERNAL:48513086-48560225\tstrand: 1, start: 0, end: 46853\n",
      "interval: chr10_MATERNAL:49784540-49831705\tstrand: -1, start: 0, end: 46853\n",
      "interval: chr10_PATERNAL:48447360-48494479\tstrand: 1, start: 0, end: 46853\n",
      "interval: chr10_PATERNAL:49717006-49764147\tstrand: -1, start: 0, end: 46853\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chr10:46514225-46561077(46853bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be checked by extend_2_left\n",
      "**Using extend_2_left to check the pos**\n",
      "extending window size 4685\n",
      "new interval chr10:46509540-46561077 51538\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr10:46504855-46561077 56223\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46500170-46561077 60908\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr10:46495485-46561077 65593\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46490800-46561077 70278\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46486115-46561077 74963\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46481430-46561077 79648\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46476745-46561077 84333\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46472060-46561077 89018\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46467375-46561077 93703\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46462690-46561077 98388\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46458005-46561077 103073\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46453320-46561077 107758\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46448635-46561077 112443\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46443950-46561077 117128\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46439265-46561077 121813\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46434580-46561077 126498\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46429895-46561077 131183\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46425210-46561077 135868\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46420525-46561077 140553\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46415840-46561077 145238\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46411155-46561077 149923\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46406470-46561077 154608\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46401785-46561077 159293\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46397100-46561077 163978\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46392415-46561077 168663\n",
      "[[0, 2], [3814, 3]]\n",
      "extend pos to the left by 117996bp\n",
      "extending window size 4685\n",
      "new interval chr10:46509540-46561077 51538\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr10:46504855-46561077 56223\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46500170-46561077 60908\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr10:46495485-46561077 65593\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46490800-46561077 70278\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr10:46486115-46561077 74963\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46481430-46561077 79648\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46476745-46561077 84333\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46472060-46561077 89018\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46467375-46561077 93703\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46462690-46561077 98388\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46458005-46561077 103073\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46453320-46561077 107758\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46448635-46561077 112443\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46443950-46561077 117128\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46439265-46561077 121813\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46434580-46561077 126498\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46429895-46561077 131183\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46425210-46561077 135868\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46420525-46561077 140553\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46415840-46561077 145238\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46411155-46561077 149923\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46406470-46561077 154608\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46401785-46561077 159293\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46397100-46561077 163978\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr10:46392415-46561077 168663\n",
      "[[0, 2], [3814, 3]]\n",
      "extend pos to the left by 117996bp\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be checked by extend_2_right\n",
      "**Using extend_2_right to check the end**\n",
      "extending window size 4685\n",
      "new interval chr10:46514225-46565763 51539\n",
      "[[51539, 5]]\n",
      "51539 100.0 %\n",
      "\n",
      "new interval chr10:46514225-46570448 56224\n",
      "[[56224, 5]]\n",
      "56224 100.0 %\n",
      "new interval chr10:46514225-46575133 60909\n",
      "[[60909, 5]]\n",
      "60909 100.0 %\n",
      "\n",
      "new interval chr10:46514225-46579818 65594\n",
      "[[65594, 5]]\n",
      "65594 100.0 %\n",
      "new interval chr10:46514225-46584503 70279\n",
      "[[70279, 5]]\n",
      "70279 100.0 %\n",
      "\n",
      "new interval chr10:46514225-46589188 74964\n",
      "[[74964, 5]]\n",
      "74964 100.0 %\n",
      "new interval chr10:46514225-46593873 79649\n",
      "[[79649, 5]]\n",
      "79649 100.0 %\n",
      "new interval chr10:46514225-46598558 84334\n",
      "[[84285, 1], [84334, 4]]\n",
      "The sequence isn't homozygous\n",
      "extend end to the right by 37431bp\n",
      "extending window size 4685\n",
      "new interval chr10:46514225-46565763 51539\n",
      "[[51539, 5]]\n",
      "51539 100.0 %\n",
      "\n",
      "new interval chr10:46514225-46570448 56224\n",
      "[[56224, 5]]\n",
      "56224 100.0 %\n",
      "new interval chr10:46514225-46575133 60909\n",
      "[[60909, 5]]\n",
      "60909 100.0 %\n",
      "\n",
      "new interval chr10:46514225-46579818 65594\n",
      "[[65594, 5]]\n",
      "65594 100.0 %\n",
      "new interval chr10:46514225-46584503 70279\n",
      "[[70279, 5]]\n",
      "70279 100.0 %\n",
      "\n",
      "new interval chr10:46514225-46589188 74964\n",
      "[[74964, 5]]\n",
      "74964 100.0 %\n",
      "new interval chr10:46514225-46593873 79649\n",
      "[[79649, 5]]\n",
      "79649 100.0 %\n",
      "new interval chr10:46514225-46598558 84334\n",
      "[[84285, 1], [84334, 4]]\n",
      "The sequence isn't homozygous\n",
      "extend end to the right by 37431bp\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chr10:46514225-46561077 (46853bp)\n",
      "The corrected interval is chr10:46396229-46598510 (202282bp)\n",
      "The corrected interval is 155429bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr10:46396229-46598510\n",
      "hg38 dup count: 1\n",
      "interval: chr10:46396228-46598510\tstrand: 1, start: 0, end: 202282\n",
      "hg2 dup count: 5\n",
      "interval: chr10_MATERNAL:47725006-47927922\tstrand: 1, start: 0, end: 202280\n",
      "interval: chr10_MATERNAL:48394468-48597666\tstrand: 1, start: 0, end: 202280\n",
      "interval: chr10_MATERNAL:49747154-49950214\tstrand: -1, start: 0, end: 202278\n",
      "interval: chr10_PATERNAL:48328957-48531912\tstrand: 1, start: 0, end: 202282\n",
      "interval: chr10_PATERNAL:49679603-49882691\tstrand: -1, start: 0, end: 202281\n",
      "input dup interval: chr10:46396229-46598510\n",
      "hg38 dup count: 1\n",
      "interval: chr10:46396228-46598510\tstrand: 1, start: 0, end: 202282\n",
      "hg2 dup count: 5\n",
      "interval: chr10_MATERNAL:47725006-47927922\tstrand: 1, start: 0, end: 202280\n",
      "interval: chr10_MATERNAL:48394468-48597666\tstrand: 1, start: 0, end: 202280\n",
      "interval: chr10_MATERNAL:49747154-49950214\tstrand: -1, start: 0, end: 202278\n",
      "interval: chr10_PATERNAL:48328957-48531912\tstrand: 1, start: 0, end: 202282\n",
      "interval: chr10_PATERNAL:49679603-49882691\tstrand: -1, start: 0, end: 202281\n"
     ]
    },
    {
     "data": {
      "text/plain": "{'dup_interval': 'chr10:46514225-46561077',\n 'dup_interval_size': 46853,\n 'original_hg38_hit_count': 1,\n 'original_hg2_hit_count': 5,\n 'original_hg2_mat_hit_count': 3,\n 'original_hg2_pat_hit_count': 2,\n 'dup_interval_classification': 'Homozygous Duplication',\n 'corrected_interval': 'chr10:46396229-46598510',\n 'corrected_interval_size': 202282,\n 'expanded_length': 155429,\n 'corrected_hg38_hit_count': 1,\n 'corrected_hg2_hit_count': 5,\n 'corrected_hg2_mat_hit_count': 3,\n 'corrected_hg2_pat_hit_count': 2,\n 'corrected_dup_interval_classification': 'Homozygous Duplication',\n 'corrected_pos_accuracy': 1,\n 'corrected_end_accuracy': 0}"
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example 6: The corrected DUP END doesn't make sense\n",
    "# Need further investigation\n",
    "duplication_evaluation.eval_dup_interval(\"chr10:46514225-46561077\", ReferenceLocations.hg38, ReferenceLocations.hg002t2t, hg38, hg002t2t).analyzeDuplicationIntervals()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-13T14:17:10.649021Z",
     "start_time": "2023-12-13T14:16:49.878399Z"
    }
   },
   "id": "e31a9840df0f9a17"
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr14:19725013-19956413\n",
      "hg38 dup count: 1\n",
      "interval: chr14:19725012-19956413\tstrand: 1, start: 0, end: 231401\n",
      "hg2 dup count: 4\n",
      "interval: chr14_MATERNAL:20204466-20435991\tstrand: 1, start: 0, end: 231401\n",
      "interval: chr14_MATERNAL:21675534-21906954\tstrand: 1, start: 0, end: 231401\n",
      "interval: chr14_PATERNAL:19420937-19652399\tstrand: 1, start: 0, end: 231401\n",
      "interval: chr14_PATERNAL:18710102-18938130\tstrand: 1, start: 0, end: 231401\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chr14:19725013-19956413(231401bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be checked by extend_2_left\n",
      "**Using extend_2_left to check the pos**\n",
      "extending window size 23140\n",
      "new interval chr14:19701873-19956413 254541\n",
      "[[0, 4]]\n",
      "0 100.0 %\n",
      "new interval chr14:19678733-19956413 277681\n",
      "[[0, 4]]\n",
      "0 100.0 %\n",
      "new interval chr14:19655593-19956413 300821\n",
      "[[0, 4]]\n",
      "0 100.0 %\n",
      "new interval chr14:19632453-19956413 323961\n",
      "[[0, 4]]\n",
      "0 100.0 %\n",
      "new interval chr14:19609313-19956413 347101\n",
      "[[2401, 4]]\n",
      "This sequence might hit the unresolved region in hg38\n",
      "extend pos to the left by 113299bp\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be checked by extend_2_right\n",
      "**Using extend_2_right to check the end**\n",
      "extending window size 23140\n",
      "new interval chr14:19725013-19979554 254542\n",
      "[[231585, 2], [254542, 2]]\n",
      "extend end to the right by 183bp\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chr14:19725013-19956413 (231401bp)\n",
      "The corrected interval is chr14:19611714-19956598 (344885bp)\n",
      "The corrected interval is 113484bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr14:19611714-19956598\n",
      "hg38 dup count: 1\n",
      "interval: chr14:19611713-19956598\tstrand: 1, start: 0, end: 344885\n",
      "hg2 dup count: 4\n",
      "interval: chr14_MATERNAL:20091276-20436164\tstrand: 1, start: 0, end: 344884\n",
      "interval: chr14_MATERNAL:21562569-21907139\tstrand: 1, start: 0, end: 344885\n",
      "interval: chr14_PATERNAL:19307728-19652584\tstrand: 1, start: 0, end: 344885\n",
      "interval: chr14_PATERNAL:18597140-18938303\tstrand: 1, start: 0, end: 344884\n",
      "----------------------------------------------------------------\n",
      "End of Analysis of DUP interval chr14:19725013-19956413(231401bp)\n",
      "----------------------------------------------------------------\n"
     ]
    },
    {
     "data": {
      "text/plain": "{'dup_interval': 'chr14:19725013-19956413',\n 'dup_interval_size': 231401,\n 'original_hg38_hit_count': 1,\n 'original_hg2_hit_count': 4,\n 'original_hg2_mat_hit_count': 2,\n 'original_hg2_pat_hit_count': 2,\n 'original_dup_interval_major_classification': 'Duplication',\n 'original_dup_interval_sub_classification': 'Homozygous Duplication',\n 'original_dup_interval_contain_big_gap': False,\n 'corrected_interval': 'chr14:19611714-19956598',\n 'corrected_interval_size': 344885,\n 'expanded_length': 113484,\n 'corrected_hg38_hit_count': 1,\n 'corrected_hg2_hit_count': 4,\n 'corrected_hg2_mat_hit_count': 2,\n 'corrected_hg2_pat_hit_count': 2,\n 'corrected_dup_interval_major_classification': 'Duplication',\n 'corrected_dup_interval_sub_classification': 'Homozygous Duplication',\n 'corrected_pos_accuracy': 0,\n 'corrected_end_accuracy': 1}"
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example 7: Basepair correction hits unresolved region (NNNNNNN) \n",
    "# For this case, the correction will end at the edge of the unresolved region\n",
    "duplication_evaluation.eval_dup_interval(\"chr14:19725013-19956413\", ReferenceLocations.hg38, ReferenceLocations.hg002t2t, hg38, hg002t2t).analyzeDuplicationIntervals()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:19:49.613684Z",
     "start_time": "2023-12-15T02:19:30.328132Z"
    }
   },
   "id": "55d73540f9a907bb"
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HG2 DEL intervals: 996\n",
      "HG2 DUP intervals: 114\n"
     ]
    }
   ],
   "source": [
    "# Perform DUP evaluation on all the HG002 DUP intervals from DRAGEN\n",
    "HG2_DRAGEN_cnv_path = \"/Users/gaoyueya/Documents/Projects/TAG-Ticket1639/DRAGEN4_2_4_visualization/DRAGEN_Output/NA24385.cnv_sv.vcf\"\n",
    "def read_vcf(vcf_path):\n",
    "    df = pd.read_csv(vcf_path, comment='#', sep='\\t', header=None, names=['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SAMPLE'])\n",
    "    return df\n",
    "HG2_DRAGEN_cnv = read_vcf(HG2_DRAGEN_cnv_path)\n",
    "HG2_DRAGEN_pass_cnv = HG2_DRAGEN_cnv[HG2_DRAGEN_cnv['FILTER']=='PASS']\n",
    "HG2_DRAGEN_pass_cnv.head()\n",
    "\n",
    "# Gather all the intervals for DEL and DUP from HG2 DRAGEN PASS\n",
    "DEL_pass_intervals = []\n",
    "DUP_pass_intervals = []\n",
    "for index, row in HG2_DRAGEN_pass_cnv.iterrows():\n",
    "    interval = row['CHROM'] + ':' + str(row['POS']) + '-' + str(row['INFO'].split('END=')[1].split(';')[0])\n",
    "    if row['ALT'] == '<DEL>':\n",
    "        DEL_pass_intervals.append(interval)\n",
    "    elif row['ALT'] == '<DUP>':\n",
    "        DUP_pass_intervals.append(interval)\n",
    "\n",
    "# Remove the DUP interval that is a known FP\n",
    "DUP_pass_intervals.remove(\"chr1:789481-224014488\") # This is a known FP that was verified by illumina\n",
    "\n",
    "print('HG2 DEL intervals:',len(DEL_pass_intervals))\n",
    "print('HG2 DUP intervals:',len(DUP_pass_intervals))"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:23:34.571096Z",
     "start_time": "2023-12-15T02:23:34.534136Z"
    }
   },
   "id": "8e662b6c71ddae93"
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1:13225212-13247040\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr1:13225212-13247040\n",
      "hg38 dup count: 1\n",
      "interval: chr1:13225211-13247040\tstrand: 1, start: 0, end: 21829\n",
      "hg2 dup count: 3\n",
      "interval: chr1_MATERNAL:12654730-12676537\tstrand: -1, start: 0, end: 21829\n",
      "interval: chr1_MATERNAL:12868220-12890084\tstrand: 1, start: 0, end: 21829\n",
      "interval: chr1_PATERNAL:12618932-12640735\tstrand: -1, start: 0, end: 21829\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chr1:13225212-13247040(21829bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be checked by extend_2_left\n",
      "**Using extend_2_left to check the pos**\n",
      "extending window size 2183\n",
      "new interval chr1:13223029-13247040 24012\n",
      "[[0, 3]]\n",
      "0 100.0 %\n",
      "new interval chr1:13220846-13247040 26195\n",
      "[[0, 2], [43, 1]]\n",
      "extend pos to the left by 4323bp\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be checked by extend_2_right\n",
      "**Using extend_2_right to check the end**\n",
      "extending window size 2183\n",
      "new interval chr1:13225212-13249224 24013\n",
      "[[24013, 3]]\n",
      "24013 100.0 %\n",
      "new interval chr1:13225212-13251407 26196\n",
      "[[26196, 3]]\n",
      "26196 100.0 %\n",
      "\n",
      "new interval chr1:13225212-13253590 28379\n",
      "[[28379, 3]]\n",
      "28379 100.0 %\n",
      "new interval chr1:13225212-13255773 30562\n",
      "[[30562, 3]]\n",
      "30562 100.0 %\n",
      "\n",
      "new interval chr1:13225212-13257956 32745\n",
      "[[32745, 3]]\n",
      "32745 100.0 %\n",
      "new interval chr1:13225212-13260139 34928\n",
      "[[34928, 3]]\n",
      "34928 100.0 %\n",
      "\n",
      "new interval chr1:13225212-13262322 37111\n",
      "[[37111, 3]]\n",
      "37111 100.0 %\n",
      "new interval chr1:13225212-13264505 39294\n",
      "[[39294, 4]]\n",
      "39294 100.0 %\n",
      "\n",
      "new interval chr1:13225212-13266688 41477\n",
      "[[41477, 4]]\n",
      "41477 100.0 %\n",
      "new interval chr1:13225212-13268871 43660\n",
      "[[43660, 4]]\n",
      "43660 100.0 %\n",
      "new interval chr1:13225212-13271054 45843\n",
      "[[45843, 4]]\n",
      "45843 100.0 %\n",
      "new interval chr1:13225212-13273237 48026\n",
      "[[48026, 4]]\n",
      "48026 100.0 %\n",
      "\n",
      "new interval chr1:13225212-13275420 50209\n",
      "[[50209, 4]]\n",
      "50209 100.0 %\n",
      "new interval chr1:13225212-13277603 52392\n",
      "[[52392, 4]]\n",
      "52392 100.0 %\n",
      "new interval chr1:13225212-13279786 54575\n",
      "[[54575, 4]]\n",
      "54575 100.0 %\n",
      "new interval chr1:13225212-13281969 56758\n",
      "[[56758, 4]]\n",
      "56758 100.0 %\n",
      "new interval chr1:13225212-13284152 58941\n",
      "[[58941, 4]]\n",
      "58941 100.0 %\n",
      "new interval chr1:13225212-13286335 61124\n",
      "[[61124, 4]]\n",
      "61124 100.0 %\n",
      "new interval chr1:13225212-13288518 63307\n",
      "[[63307, 4]]\n",
      "63307 100.0 %\n",
      "new interval chr1:13225212-13290701 65490\n",
      "[[65490, 4]]\n",
      "65490 100.0 %\n",
      "new interval chr1:13225212-13292884 67673\n",
      "[[67673, 4]]\n",
      "67673 100.0 %\n",
      "new interval chr1:13225212-13295067 69856\n",
      "[[69856, 4]]\n",
      "69856 100.0 %\n",
      "new interval chr1:13225212-13297250 72039\n",
      "[[72039, 4]]\n",
      "72039 100.0 %\n",
      "new interval chr1:13225212-13299433 74222\n",
      "[[74222, 4]]\n",
      "74222 100.0 %\n",
      "new interval chr1:13225212-13301616 76405\n",
      "[[76405, 4]]\n",
      "76405 100.0 %\n",
      "new interval chr1:13225212-13303799 78588\n",
      "[[78588, 4]]\n",
      "78588 100.0 %\n",
      "new interval chr1:13225212-13305982 80771\n",
      "[[80771, 4]]\n",
      "80771 100.0 %\n",
      "new interval chr1:13225212-13308165 82954\n",
      "[[82954, 4]]\n",
      "82954 100.0 %\n",
      "new interval chr1:13225212-13310348 85137\n",
      "[[85137, 4]]\n",
      "85137 100.0 %\n",
      "new interval chr1:13225212-13312531 87320\n",
      "[[87320, 4]]\n",
      "87320 100.0 %\n",
      "new interval chr1:13225212-13314714 89503\n",
      "[[89503, 4]]\n",
      "89503 100.0 %\n",
      "new interval chr1:13225212-13316897 91686\n",
      "[[91686, 4]]\n",
      "91686 100.0 %\n",
      "new interval chr1:13225212-13319080 93869\n",
      "[[93869, 4]]\n",
      "93869 100.0 %\n",
      "new interval chr1:13225212-13321263 96052\n",
      "[[96052, 4]]\n",
      "96052 100.0 %\n",
      "new interval chr1:13225212-13323446 98235\n",
      "[[98235, 4]]\n",
      "98235 100.0 %\n",
      "new interval chr1:13225212-13325629 100418\n",
      "[[100418, 4]]\n",
      "100418 100.0 %\n",
      "new interval chr1:13225212-13327812 102601\n",
      "[[102285, 2], [102601, 2]]\n",
      "extend end to the right by 80455bp\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chr1:13225212-13247040 (21829bp)\n",
      "The corrected interval is chr1:13220889-13327497 (106609bp)\n",
      "The corrected interval is 84780bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr1:13220889-13327497\n",
      "hg38 dup count: 1\n",
      "interval: chr1:13220888-13327497\tstrand: 1, start: 0, end: 106609\n",
      "hg2 dup count: 4\n",
      "interval: chr1_MATERNAL:12863835-12970548\tstrand: 1, start: 0, end: 106609\n",
      "interval: chr1_MATERNAL:12574323-12680861\tstrand: -1, start: 0, end: 106608\n",
      "interval: chr1_PATERNAL:12538521-12645059\tstrand: -1, start: 0, end: 106608\n",
      "interval: chr1_PATERNAL:12787053-12886853\tstrand: 1, start: 1485, end: 106609\n",
      "----------------------------------------------------------------\n",
      "End of Analysis of DUP interval chr1:13225212-13247040(21829bp)\n",
      "----------------------------------------------------------------\n",
      "chr1:16605768-16645359\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Original Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr1:16605768-16645359\n",
      "hg38 dup count: 1\n",
      "interval: chr1:16605767-16645359\tstrand: 1, start: 0, end: 39592\n",
      "hg2 dup count: 5\n",
      "interval: chr1_MATERNAL:16887164-16926759\tstrand: -1, start: 0, end: 39592\n",
      "interval: chr1_MATERNAL:16556347-16595891\tstrand: 1, start: 0, end: 39592\n",
      "interval: chr1_PATERNAL:16899366-16938979\tstrand: -1, start: 0, end: 39592\n",
      "interval: chr1_PATERNAL:16376753-16416341\tstrand: 1, start: 0, end: 39592\n",
      "interval: chr1_PATERNAL:16716679-16756269\tstrand: -1, start: 0, end: 39592\n",
      "----------------------------------------------------------------\n",
      "Analyzing the DUP interval chr1:16605768-16645359(39592bp)\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE POS OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval pos needs to be checked by extend_2_left\n",
      "**Using extend_2_left to check the pos**\n",
      "extending window size 3959\n",
      "new interval chr1:16601809-16645359 43551\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr1:16597850-16645359 47510\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16593891-16645359 51469\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "\n",
      "new interval chr1:16589932-16645359 55428\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16585973-16645359 59387\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16582014-16645359 63346\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16578055-16645359 67305\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16574096-16645359 71264\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16570137-16645359 75223\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16566178-16645359 79182\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16562219-16645359 83141\n",
      "[[0, 4], [1509, 1]]\n",
      "1509 100.0 %\n",
      "new interval chr1:16558260-16645359 87100\n",
      "[[0, 5]]\n",
      "0 100.0 %\n",
      "new interval chr1:16554301-16645359 91059\n",
      "[[0, 6]]\n",
      "0 100.0 %\n",
      "new interval chr1:16550342-16645359 95018\n",
      "[[0, 3], [2528, 3]]\n",
      "2528 100.0 %\n",
      "new interval chr1:16546383-16645359 98977\n",
      "[[0, 3], [6487, 3]]\n",
      "6487 100.0 %\n",
      "new interval chr1:16542424-16645359 102936\n",
      "[[0, 3], [10446, 3]]\n",
      "10446 100.0 %\n",
      "new interval chr1:16538465-16645359 106895\n",
      "[[0, 4], [14405, 2]]\n",
      "The sequence isn't homozygous\n",
      "extend pos to the left by 52898bp\n",
      "----------------------------------------------------------------\n",
      "ANALYZING THE END OF THE DUP INTERVAL\n",
      "----------------------------------------------------------------\n",
      "The called interval end needs to be checked by extend_2_right\n",
      "**Using extend_2_right to check the end**\n",
      "extending window size 3959\n",
      "new interval chr1:16605768-16649319 43552\n",
      "[[43552, 5]]\n",
      "43552 100.0 %\n",
      "\n",
      "new interval chr1:16605768-16653278 47511\n",
      "[[47511, 5]]\n",
      "47511 100.0 %\n",
      "new interval chr1:16605768-16657237 51470\n",
      "[[51470, 5]]\n",
      "51470 100.0 %\n",
      "\n",
      "new interval chr1:16605768-16661196 55429\n",
      "[[55120, 1], [55429, 4]]\n",
      "The sequence isn't homozygous\n",
      "extend end to the right by 15527bp\n",
      "----------------------------------------------------------------\n",
      "Corrected Interval Details\n",
      "----------------------------------------------------------------\n",
      "The original interval is chr1:16605768-16645359 (39592bp)\n",
      "The corrected interval is chr1:16552870-16660888 (108019bp)\n",
      "The corrected interval is 68427bp longer than the original interval\n",
      "----------------------------------------------------------------\n",
      "Alignment Report of Corrected Interval\n",
      "----------------------------------------------------------------\n",
      "input dup interval: chr1:16552870-16660888\n",
      "hg38 dup count: 1\n",
      "interval: chr1:16552869-16660888\tstrand: 1, start: 0, end: 108019\n",
      "hg2 dup count: 5\n",
      "interval: chr1_MATERNAL:16871633-16981196\tstrand: -1, start: 0, end: 108019\n",
      "interval: chr1_MATERNAL:16496938-16611419\tstrand: 1, start: 0, end: 108019\n",
      "interval: chr1_PATERNAL:16883876-17012255\tstrand: -1, start: 0, end: 108018\n",
      "interval: chr1_PATERNAL:16322107-16431867\tstrand: 1, start: 0, end: 108019\n",
      "interval: chr1_PATERNAL:16701153-16820407\tstrand: -1, start: 0, end: 108019\n",
      "----------------------------------------------------------------\n",
      "End of Analysis of DUP interval chr1:16605768-16645359(39592bp)\n",
      "----------------------------------------------------------------\n"
     ]
    },
    {
     "data": {
      "text/plain": "                 interval  dup_interval_size  original_hg38_hit_count  \\\n0  chr1:13225212-13247040              21829                        1   \n1  chr1:16605768-16645359              39592                        1   \n\n   original_hg2_hit_count  original_hg2_mat_hit_count  \\\n0                       3                           2   \n1                       5                           2   \n\n   original_hg2_pat_hit_count original_dup_interval_major_classification  \\\n0                           1                                Duplication   \n1                           3                                Duplication   \n\n  original_dup_interval_sub_classification  \\\n0        Maternal Heterozygous Duplication   \n1                   Homozygous Duplication   \n\n   original_dup_interval_contain_big_gap      corrected_interval  \\\n0                                  False  chr1:13220889-13327497   \n1                                  False  chr1:16552870-16660888   \n\n   corrected_interval_size  expanded_length  corrected_hg38_hit_count  \\\n0                   106609            84780                         1   \n1                   108019            68427                         1   \n\n   corrected_hg2_hit_count  corrected_hg2_mat_hit_count  \\\n0                        4                            2   \n1                        5                            2   \n\n   corrected_hg2_pat_hit_count corrected_dup_interval_major_classification  \\\n0                            2                                 Duplication   \n1                            3                                 Duplication   \n\n  corrected_dup_interval_sub_classification  corrected_pos_accuracy  \\\n0                    Homozygous Duplication                       1   \n1                    Homozygous Duplication                       0   \n\n   corrected_end_accuracy  \n0                       1  \n1                       0  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>interval</th>\n      <th>dup_interval_size</th>\n      <th>original_hg38_hit_count</th>\n      <th>original_hg2_hit_count</th>\n      <th>original_hg2_mat_hit_count</th>\n      <th>original_hg2_pat_hit_count</th>\n      <th>original_dup_interval_major_classification</th>\n      <th>original_dup_interval_sub_classification</th>\n      <th>original_dup_interval_contain_big_gap</th>\n      <th>corrected_interval</th>\n      <th>corrected_interval_size</th>\n      <th>expanded_length</th>\n      <th>corrected_hg38_hit_count</th>\n      <th>corrected_hg2_hit_count</th>\n      <th>corrected_hg2_mat_hit_count</th>\n      <th>corrected_hg2_pat_hit_count</th>\n      <th>corrected_dup_interval_major_classification</th>\n      <th>corrected_dup_interval_sub_classification</th>\n      <th>corrected_pos_accuracy</th>\n      <th>corrected_end_accuracy</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>chr1:13225212-13247040</td>\n      <td>21829</td>\n      <td>1</td>\n      <td>3</td>\n      <td>2</td>\n      <td>1</td>\n      <td>Duplication</td>\n      <td>Maternal Heterozygous Duplication</td>\n      <td>False</td>\n      <td>chr1:13220889-13327497</td>\n      <td>106609</td>\n      <td>84780</td>\n      <td>1</td>\n      <td>4</td>\n      <td>2</td>\n      <td>2</td>\n      <td>Duplication</td>\n      <td>Homozygous Duplication</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>chr1:16605768-16645359</td>\n      <td>39592</td>\n      <td>1</td>\n      <td>5</td>\n      <td>2</td>\n      <td>3</td>\n      <td>Duplication</td>\n      <td>Homozygous Duplication</td>\n      <td>False</td>\n      <td>chr1:16552870-16660888</td>\n      <td>108019</td>\n      <td>68427</td>\n      <td>1</td>\n      <td>5</td>\n      <td>2</td>\n      <td>3</td>\n      <td>Duplication</td>\n      <td>Homozygous Duplication</td>\n      <td>0</td>\n      <td>0</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg2_dup_summary_df = pd.DataFrame()\n",
    "hg2_dup_summary_df['interval'] = DUP_pass_intervals[:2]\n",
    "\n",
    "# Collect the results for each interval from dup_eval\n",
    "dup_interval_size_list = []\n",
    "original_hg38_hit_count_list = []\n",
    "original_hg2_hit_count_list = []\n",
    "original_hg2_mat_hit_count_list = []\n",
    "original_hg2_pat_hit_count_list = []\n",
    "original_dup_interval_major_classification_list = []\n",
    "original_dup_interval_sub_classification_list = []\n",
    "original_dup_interval_contain_big_gap_list = []\n",
    "corrected_interval_list = []\n",
    "corrected_interval_size_list = []\n",
    "expanded_length_list = []\n",
    "corrected_hg38_hit_count_list = []\n",
    "corrected_hg2_hit_count_list = []\n",
    "corrected_hg2_mat_hit_count_list = []\n",
    "corrected_hg2_pat_hit_count_list = []\n",
    "corrected_dup_interval_major_classification_list = []\n",
    "corrected_dup_interval_sub_classification_list = []\n",
    "corrected_pos_accuracy_list = []\n",
    "corrected_end_accuracy_list = []\n",
    "\n",
    "for interval in DUP_pass_intervals[:2]:\n",
    "    print(interval)\n",
    "    interval_sum_dict = duplication_evaluation.eval_dup_interval(interval, ReferenceLocations.hg38, ReferenceLocations.hg002t2t, hg38, hg002t2t).analyzeDuplicationIntervals()\n",
    "    dup_interval_size_list.append(interval_sum_dict['dup_interval_size'])\n",
    "    original_hg38_hit_count_list.append(interval_sum_dict['original_hg38_hit_count'])\n",
    "    original_hg2_hit_count_list.append(interval_sum_dict['original_hg2_hit_count'])\n",
    "    original_hg2_mat_hit_count_list.append(interval_sum_dict['original_hg2_mat_hit_count'])\n",
    "    original_hg2_pat_hit_count_list.append(interval_sum_dict['original_hg2_pat_hit_count'])\n",
    "    original_dup_interval_major_classification_list.append(interval_sum_dict['original_dup_interval_major_classification'])\n",
    "    original_dup_interval_sub_classification_list.append(interval_sum_dict['original_dup_interval_sub_classification'])\n",
    "    original_dup_interval_contain_big_gap_list.append(interval_sum_dict['original_dup_interval_contain_big_gap'])\n",
    "    corrected_interval_list.append(interval_sum_dict['corrected_interval'])\n",
    "    corrected_interval_size_list.append(interval_sum_dict['corrected_interval_size'])\n",
    "    expanded_length_list.append(interval_sum_dict['expanded_length'])\n",
    "    corrected_hg38_hit_count_list.append(interval_sum_dict['corrected_hg38_hit_count'])\n",
    "    corrected_hg2_hit_count_list.append(interval_sum_dict['corrected_hg2_hit_count'])\n",
    "    corrected_hg2_mat_hit_count_list.append(interval_sum_dict['corrected_hg2_mat_hit_count'])\n",
    "    corrected_hg2_pat_hit_count_list.append(interval_sum_dict['corrected_hg2_pat_hit_count'])\n",
    "    corrected_dup_interval_major_classification_list.append(interval_sum_dict['corrected_dup_interval_major_classification'])\n",
    "    corrected_dup_interval_sub_classification_list.append(interval_sum_dict['corrected_dup_interval_sub_classification'])\n",
    "    corrected_pos_accuracy_list.append(interval_sum_dict['corrected_pos_accuracy'])\n",
    "    corrected_end_accuracy_list.append(interval_sum_dict['corrected_end_accuracy'])\n",
    "    \n",
    "hg2_dup_summary_df['dup_interval_size'] = dup_interval_size_list\n",
    "hg2_dup_summary_df['original_hg38_hit_count'] = original_hg38_hit_count_list\n",
    "hg2_dup_summary_df['original_hg2_hit_count'] = original_hg2_hit_count_list\n",
    "hg2_dup_summary_df['original_hg2_mat_hit_count'] = original_hg2_mat_hit_count_list\n",
    "hg2_dup_summary_df['original_hg2_pat_hit_count'] = original_hg2_pat_hit_count_list\n",
    "hg2_dup_summary_df['original_dup_interval_major_classification'] = original_dup_interval_major_classification_list\n",
    "hg2_dup_summary_df['original_dup_interval_sub_classification'] = original_dup_interval_sub_classification_list\n",
    "hg2_dup_summary_df['original_dup_interval_contain_big_gap'] = original_dup_interval_contain_big_gap_list\n",
    "hg2_dup_summary_df['corrected_interval'] = corrected_interval_list\n",
    "hg2_dup_summary_df['corrected_interval_size'] = corrected_interval_size_list\n",
    "hg2_dup_summary_df['expanded_length'] = expanded_length_list\n",
    "hg2_dup_summary_df['corrected_hg38_hit_count'] = corrected_hg38_hit_count_list\n",
    "hg2_dup_summary_df['corrected_hg2_hit_count'] = corrected_hg2_hit_count_list\n",
    "hg2_dup_summary_df['corrected_hg2_mat_hit_count'] = corrected_hg2_mat_hit_count_list\n",
    "hg2_dup_summary_df['corrected_hg2_pat_hit_count'] = corrected_hg2_pat_hit_count_list\n",
    "hg2_dup_summary_df['corrected_dup_interval_major_classification'] = corrected_dup_interval_major_classification_list\n",
    "hg2_dup_summary_df['corrected_dup_interval_sub_classification'] = corrected_dup_interval_sub_classification_list\n",
    "hg2_dup_summary_df['corrected_pos_accuracy'] = corrected_pos_accuracy_list\n",
    "hg2_dup_summary_df['corrected_end_accuracy'] = corrected_end_accuracy_list\n",
    "\n",
    "hg2_dup_summary_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-12-15T02:32:24.563250Z",
     "start_time": "2023-12-15T02:32:02.448725Z"
    }
   },
   "id": "1f1d0a0cebe8ac3d"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "9c975c14bd021914"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
